src/stream/src/executor/over_window/general.rs

```diff
@@ -0,0 +1,921 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::marker::PhantomData;
+use std::ops::Bound;
+
+use futures::StreamExt;
+use futures_async_stream::{for_await, try_stream};
+use itertools::Itertools;
+use risingwave_common::array::stream_record::Record;
+use risingwave_common::array::{RowRef, StreamChunk};
+use risingwave_common::catalog::Field;
+use risingwave_common::row::{OwnedRow, Row, RowExt};
+use risingwave_common::types::DefaultOrdered;
+use risingwave_common::util::iter_util::ZipEqFast;
+use risingwave_common::util::memcmp_encoding::{self, MemcmpEncoded};
+use risingwave_common::util::sort_util::OrderType;
+use risingwave_expr::function::window::{FrameBounds, WindowFuncCall};
+use risingwave_storage::store::PrefetchOptions;
+use risingwave_storage::StateStore;
+
+use self::private::Partition;
+use super::diff_btree_map::{Change, DiffBTreeMap};
+use super::state::{create_window_state, StateKey};
+use crate::cache::{new_unbounded, ManagedLruCache};
+use crate::common::metrics::MetricsInfo;
+use crate::executor::aggregation::ChunkBuilder;
+use crate::executor::over_window::diff_btree_map::PositionType;
+use crate::executor::over_window::window_states::WindowStates;
+use crate::executor::test_utils::prelude::StateTable;
+use crate::executor::{
+    expect_first_barrier, ActorContextRef, BoxedExecutor, Executor, ExecutorInfo, Message,
+    PkIndices, StreamExecutorError, StreamExecutorResult,
+};
+use crate::task::AtomicU64Ref;
+
+mod private {
+    use std::collections::BTreeMap;
+
+    use risingwave_common::estimate_size::EstimateSize;
+    use risingwave_common::row::OwnedRow;
+
+    use crate::executor::over_window::state::StateKey;
+
+    pub(super) struct Partition {
+        /// Fully synced table cache for the partition. `StateKey (order key, input pk)` -> table
+        /// row.
+        cache: BTreeMap<StateKey, OwnedRow>,
+        heap_size: usize,
```

Comment:
Use this for heap_size. https://github.com/singularity-data/risingwave/blob/6b4b4e657b83d06472699e5dfb0db0a39b5f852c/src/common/src/estimate_size/mod.rs#L190